modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/config/DynamicConfigurationInstanceProvider.java

```diff
@@ -0,0 +1,180 @@
+            operationContext.onOperationFailed(signal -> discountUsage(configurationInstanceHolder));
+
+            configurationInstanceHolder.accountUsage();
+            return (T) configurationInstanceHolder.getConfigurationInstance();
+        }
+        catch (Exception e)
+        {
+            throw new MuleRuntimeException(e);
+        }
+    }
+
+    private DynamicConfigurationInstanceHolder getHolder(ResolverSetResult result) throws Exception
+    {
+        DynamicConfigurationInstanceHolder configurationInstance;
+        cacheReadLock.lock();
+        try
+        {
+            configurationInstance = cache.get(result);
+            if (configurationInstance == null)
+            {
+                cacheReadLock.unlock();
+                cacheWriteLock.lock();
+
+                try
+                {
+                    // re-check in case some other thread beat us to it...
+                    configurationInstance = cache.get(result);
+                    if (configurationInstance == null)
+                    {
+                        configurationInstance = createAndRegisterConfigurationInstance(result);
+                        cache.put(result, configurationInstance);
+                    }
+
+                    // downgrade
+                    cacheReadLock.lock();
+                }
+                finally
+                {
+                    cacheWriteLock.unlock();
+                }
+            }
+        }
+        finally
+        {
+            cacheReadLock.unlock();
+        }
+
+        return configurationInstance;
+    }
+
+    private DynamicConfigurationInstanceHolder createAndRegisterConfigurationInstance(ResolverSetResult result) throws MuleException
+    {
+        Object configurationInstance = configurationObjectBuilder.build(result);
+        String registrationName = registrationCallback.registerConfigurationInstance(extension, name, configurationInstance);
+        return new DynamicConfigurationInstanceHolder(registrationName, configurationInstance);
+    }
+
+    @Override
+    public Map<String, Object> getExpired()
+    {
+        ImmutableMap.Builder<String, Object> expiredConfigs = ImmutableMap.builder();
+        cacheWriteLock.lock();
+        try
+        {
+            cache.entrySet()
+                    .stream()
+                    .filter(config -> config.getValue().isExpired(expirationPolicy))
+                    .forEach(config -> {
+                        cache.remove(config.getKey());
+                        DynamicConfigurationInstanceHolder wrapper = config.getValue();
+                        expiredConfigs.put(wrapper.getRegistrationName(), wrapper.getConfigurationInstance());
+                    });
+        }
+        finally
+        {
+            cacheWriteLock.unlock();
+        }
+
+        return expiredConfigs.build();
+    }
+
+    private void discountUsage(DynamicConfigurationInstanceHolder configurationInstanceHolder)
```

Comment:
that's only a good idea if the two lambdas from your previous comment could be reused. Since that's not the case, I'd rather not inline this